### ISSUE TITLE
Weird packet sent from moolticute

### ISSUE THREAD

**Issue by limpkin** [mooltipass/moolticute#486]
This is a very weird one...

moolticute log: 

DEBUG: :0 - "Received answer: GET_CUR_CARD_CPZ (0x000b)"  Full packet: "0c000b0xxxxxxxxxxxxx(hidden)00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000"
DEBUG: :0 - Card CPZ:  "<<hidden>>"
INFO: :0 - Finished loading card CPZ
INFO: :0 - "Loading device db change numbers"
DEBUG: :0 - Platform send command:  "GET_USER_CHANGE_NB (0x000a)"
DEBUG: :0 - Full packet# 0 :  "[0x04, 0x00, 0x0a, 0x00, 0x00, 0x00]"
DEBUG: :0 - JSON API recv: {
    "msg": "get_user_categories"

on the usb capture i can correctly see the get_card_cpz and its answer.... but I do not see the get_user_changer_nb packet at all. Instead, I see this one:

![image](https://user-images.githubusercontent.com/6170291/68416593-0db5e200-0195-11ea-9c61-9a2e7cac6de0.png)

which is an invalid packet....

**Comment by deXol** [comment]
Is it happened on Windows?
It looks really weird, cannot really find any place where it was possible sent from MC.
What has happened after this message, is the GET_USER_CHANGE_NB messge not written at all?
And where you able to reproduce it since then?

**Comment by limpkin** [comment]
windows indeed.
it is fairly reproduceable yes.... don't know about the get_user message

**Comment by limpkin** [comment]
@deXol there's definitely something odd happening on the low level side of things.... how are buffers allocated? that looks like some buffer misuse.

This other packet is from a capture taken today:

![image](https://user-images.githubusercontent.com/6170291/69093725-513f0480-0a4f-11ea-8842-895065202ae8.png)

how come our websocket messages go to USB? this is high priority @deXol

**Comment by raoulh** [comment]
That smells like a pointer to a wrong memory space, or a QByteArray that is uninitialized. I would advise @deXol to build moolticute using ASAN. To do so you can run qmake with:
```
qmake "CONFIG+=sanitizer sanitize_address" moolticute.pro
```
other sanitzers are also available:

- `sanitize_address`
- `sanitize_memory`
- `sanitize_thread`
- `sanitize_undefined`

**Comment by deXol** [comment]
@raoulh It looks like sanitizer is not available on Windows: https://stackoverflow.com/questions/31144000/mingw-w64s-gcc-and-address-sanitizer/31151497
I tried on Linux and for GUI I was able to receive some report (LeakSanitizer: detected memory leaks), when I quit from GUI.
But for daemon it is only possible to terminate, in that case no report displayed from the Sanitizer. I tried a workaround with adding a timer and calling quit from AppDaemon, but unfortunately Sanitizer is only displayed a crash:
![image](https://user-images.githubusercontent.com/11043249/69182996-5fa22480-0b12-11ea-87c0-9c62817fa24e.png)
Do you have any idea to get the report for daemon too?

**Comment by deXol** [comment]
On windows I tried to print every QByteArray right before writing to the device, but I have never seen a message like these.
@limpkin Is there any special circumstance to reproduce this? when it is happening, right after starting daemon?

**Comment by limpkin** [comment]
It can happen at any time....

**Comment by limpkin** [comment]
@deXol could you implement #505 so i can correlate usb logs with moolticute ones?

**Comment by limpkin** [comment]
Reproduce issue:

USB trace:
- correct exchange for get device language
- correct exchange for get keyboard layout for USB
... then this:

![image](https://user-images.githubusercontent.com/6170291/69321795-033f2200-0c44-11ea-9b8e-3588330a18db.png)

![image](https://user-images.githubusercontent.com/6170291/69321734-ea367100-0c43-11ea-9694-305d0f300140.png)

Logs (relevant part only):

INFO: (2019-11-21T09:22:47.484) :0 - "Loading device parameters"
DEBUG: (2019-11-21T09:22:47.484) :0 - Platform send command:  "GET_DEVICE_SETTINGS (0x000c)"
DEBUG: (2019-11-21T09:22:47.484) :0 - Full packet# 0 :  "[0x84, 0x00, 0x0c, 0x00, 0x00, 0x00]"
DEBUG: (2019-11-21T09:22:47.488) :0 - Message payload length: 64
DEBUG: (2019-11-21T09:22:47.488) :0 - "Received answer: GET_DEVICE_SETTINGS (0x000c)"  Full packet: "3e010c004000ff0000ff000000050190000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000"
DEBUG: (2019-11-21T09:22:47.489) :0 - "Received answer: GET_DEVICE_SETTINGS (0x000c)"  Full packet: "06110000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000"
DEBUG: (2019-11-21T09:22:47.489) :0 - Full device settings payload:  "ff0000ff000000050190000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000"
DEBUG: (2019-11-21T09:22:47.490) :0 - Platform send command:  "GET_USER_LANG (0x0016)"
DEBUG: (2019-11-21T09:22:47.492) :0 - Full packet# 0 :  "[0x04, 0x00, 0x16, 0x00, 0x00, 0x00]"
DEBUG: (2019-11-21T09:22:47.495) :0 - Message payload length: 1
DEBUG: (2019-11-21T09:22:47.495) :0 - "Received answer: GET_USER_LANG (0x0016)"  Full packet: "05001600010000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000"
DEBUG: (2019-11-21T09:22:47.496) :0 - User language:  0
DEBUG: (2019-11-21T09:22:47.496) :0 - Platform send command:  "GET_DEVICE_LANG (0x0017)"
DEBUG: (2019-11-21T09:22:47.496) :0 - Full packet# 0 :  "[0x84, 0x00, 0x17, 0x00, 0x00, 0x00]"
DEBUG: (2019-11-21T09:22:47.499) :0 - Message payload length: 1
DEBUG: (2019-11-21T09:22:47.499) :0 - "Received answer: GET_DEVICE_LANG (0x0017)"  Full packet: "05001700010000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000"
DEBUG: (2019-11-21T09:22:47.499) :0 - Device language:  0
DEBUG: (2019-11-21T09:22:47.500) :0 - Platform send command:  "GET_KEYB_LAYOUT_ID (0x0018)"
DEBUG: (2019-11-21T09:22:47.501) :0 - Full packet# 0 :  "[0x05, 0x00, 0x18, 0x00, 0x01, 0x00, 0x01]"
DEBUG: (2019-11-21T09:22:47.505) :0 - Message payload length: 1
DEBUG: (2019-11-21T09:22:47.506) :0 - "Received answer: GET_KEYB_LAYOUT_ID (0x0018)"  Full packet: "05001800010016000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000"
DEBUG: (2019-11-21T09:22:47.537) :0 - Keyboard USB layout:  22
DEBUG: (2019-11-21T09:22:47.590) :0 - Platform send command:  "GET_KEYB_LAYOUT_ID (0x0018)"
DEBUG: (2019-11-21T09:22:47.676) :0 - Full packet# 0 :  "[0x85, 0x00, 0x18, 0x00, 0x01, 0x00, 0x00]"
INFO: (2019-11-21T09:22:47.694) :0 - Requesting change numbers
DEBUG: (2019-11-21T09:22:47.704) :0 - JSON API recv: {
    "msg": "get_user_categories"
}

DEBUG: (2019-11-21T09:22:55.172) :0 - JSON API recv: {
    "data": {
        "want_data": false
    },
    "msg": "start_memorymgmt"
}

**Comment by limpkin** [comment]
Another trace, extremely similar:
- correct exchange for get device language
- correct exchange for get keyboard layout for USB
- correct exchange for get keyboard layout for BLE
... then bad things

![image](https://user-images.githubusercontent.com/6170291/69328682-f6283000-0c4f-11ea-887d-113926fff7dd.png)

![image](https://user-images.githubusercontent.com/6170291/69328700-fd4f3e00-0c4f-11ea-9e40-584f4ff59aba.png)

Logs (relevant part only):

DEBUG: (2019-11-21T09:31:31.044) :0 - Platform send command:  "GET_KEYB_LAYOUT_ID (0x0018)"
DEBUG: (2019-11-21T09:31:31.044) :0 - Full packet# 0 :  "[0x05, 0x00, 0x18, 0x00, 0x01, 0x00, 0x01]"
DEBUG: (2019-11-21T09:31:31.048) :0 - Message payload length: 1
DEBUG: (2019-11-21T09:31:31.049) :0 - "Received answer: GET_KEYB_LAYOUT_ID (0x0018)"  Full packet: "05001800010016000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000"
DEBUG: (2019-11-21T09:31:31.049) :0 - Keyboard USB layout:  22
DEBUG: (2019-11-21T09:31:31.050) :0 - Platform send command:  "GET_KEYB_LAYOUT_ID (0x0018)"
DEBUG: (2019-11-21T09:31:31.050) :0 - Full packet# 0 :  "[0x85, 0x00, 0x18, 0x00, 0x01, 0x00, 0x00]"
DEBUG: (2019-11-21T09:31:31.071) :0 - Message payload length: 1
DEBUG: (2019-11-21T09:31:31.072) :0 - "Received answer: GET_KEYB_LAYOUT_ID (0x0018)"  Full packet: "05001800010000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000"
DEBUG: (2019-11-21T09:31:31.072) :0 - Keyboard Bluetooth layout:  0
INFO: (2019-11-21T09:31:31.072) :0 - "Send date to device"
DEBUG: (2019-11-21T09:31:31.073) :0 - Sending current date:  "e3070b00150009001f001f00"
DEBUG: (2019-11-21T09:31:31.074) :0 - Platform send command:  "SET_DATE (0x0004)"
DEBUG: (2019-11-21T09:31:31.075) :0 - Full packet# 0 :  "[0x10, 0x00, 0x04, 0x00, 0x0c, 0x00, 0xe3, 0x07, 0x0b, 0x00, 0x15, 0x00, 0x09, 0x00, 0x1f, 0x00, 0x1f, 0x00]"
INFO: (2019-11-21T09:31:31.075) :0 - Requesting change numbers
DEBUG: (2019-11-21T09:31:31.100) :0 - JSON API recv: {
    "msg": "get_user_categories"
}

DEBUG: (2019-11-21T09:31:37.366) :0 - JSON API recv: {
    "data": {
        "want_data": false
    },
    "msg": "start_memorymgmt"
}

Is it possible the log window is sharing a buffer with the usb sending part?

**Comment by deXol** [comment]
**Analysis:**
I think I was able to **reproduce** the issue on a **win7 virtual image**.
I cannot confirm this, because I do not have the tool for the usb packet analyzer, but I experienced the same daemon stuck exactly at the same situations.
First I was suspicious to the introduced write buffer (https://github.com/mooltipass/moolticute/pull/473/commits/8f3aba839cfb3d360952e9533e3d39716860fe16), but I tested it and working correctly, and also when I removed it the issue remained.
Looks like after the fileapi WriteFile function call (which is async) if a websocket message is received (mostly from GUI) and writing the output buffer it also ruins the written bytearray to the device. I am checking the fileapi documentation what can cause this. (https://docs.microsoft.com/en-us/windows/win32/api/fileapi/nf-fileapi-writefile)
I was also able to **reproduce** the issue **on the standard mini** too, but the retry logic fixed the communication.
I will add extra debug logs to the writeFile logic and create a PR for it.
Also @limpkin can we ask system information (os with version, is it a virtual env, CPU, memory) where the issue occured?

**Comment by limpkin** [comment]
pinging @Timandrias for systems specs

**Comment by limpkin** [comment]
This also freezes communications with the mooltipass mini standard in a very special case...

EBUG: (2019-11-23T23:19:56.991) :0 - Platform send command:  "MOOLTIPASS_STATUS (0x00b9)"
DEBUG: (2019-11-23T23:19:56.991) :0 - Full packet# 0 :  "[0x00, 0xb9]"
DEBUG: (2019-11-23T23:19:59.989) :0 - > Retry command:  "MOOLTIPASS_STATUS (0x00b9)"
DEBUG: (2019-11-23T23:20:00.111) :0 - "PLEASE_RETRY (0x00c4)"  was received for a packet that was sent due to a timeout, not resending
DEBUG: (2019-11-23T23:20:00.112) :0 - "PLEASE_RETRY (0x00c4)"  was received for a packet that was sent due to a timeout, not resending
DEBUG: (2019-11-23T23:23:59.334) :0 - JSON API recv: {
    "msg": "get_random_numbers"
}

**Comment by deXol** [comment]
I think I found the root cause.
Because `WriteFile` is async it is not guaranteed that it will happen instantly (e.g.: on my slower virtual environment). But for `WriteFile` we are passing a const char* to a local QByteArray. https://github.com/mooltipass/moolticute/blob/master/src/MPDevice_win.cpp#L195
So it is possible the program returned from the function, the local QByteArray is destructed and it has not been written to the device yet. Which would also explain the unitialized buffer written to the device.
I tried with creating a new char array, memcpy the local QByteArray and looks like the issue does not occur then.
